waiter/src/waiter/marathon_api.clj

```diff
@@ -0,0 +1,145 @@
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.marathon-api
+  (:require [clojure.core.async :as async]
+            [clojure.data.json :as json]
+            [clojure.walk :as walk]
+            [qbits.jet.client.http :as http]
+            [slingshot.slingshot :as ss]
+            [waiter.auth.spnego :as spnego])
+  (:import java.net.URI
+           org.eclipse.jetty.client.HttpClient))
+
+(defn http-request
+  "Wrapper over the qbits.jet.client.http/request function.
+   It performs a blocking read on the response and the response body.
+   The body is assumed to be json and is parsed into a clojure data structure.
+   If the status of the response in not 2XX, the response is thrown as an exception."
+  [http-client request-url & {:keys [accept body content-type headers query-string request-method
+                                     spnego-auth throw-exceptions]
+                              :or {spnego-auth false
+                                   throw-exceptions true}}]
+  (let [request-map (cond-> {:as :string
+                             :method (or request-method :get)
+                             :url request-url}
+                            spnego-auth
+                            (assoc :auth (spnego/spnego-authentication (URI. request-url)))
+                            accept
+                            (assoc :accept accept)
+                            body
+                            (assoc :body body)
+                            content-type
+                            (assoc :content-type content-type)
+                            (seq headers)
+                            (assoc :headers headers)
+                            query-string
+                            (assoc :query-string query-string))
+        raw-response (http/request http-client request-map)
+        {:keys [error status] :as response} (async/<!! raw-response)]
+    (when error
+      (throw error))
+    (when (and throw-exceptions (not (<= 200 status 299)))
+      (ss/throw+ response))
+    (-> response
+        :body
+        async/<!!
+        json/read-str
+        walk/keywordize-keys)))
+
+(defrecord MarathonApi [^HttpClient http-client ^String marathon-url spnego-auth])
+
+(defn marathon-rest-api-factory
+  "Factory method for MarathonApi."
+  [http-options url]
+  (let [http-client (http/client {:connect-timeout (:conn-timeout http-options)
+                                  :idle-timeout (:socket-timeout http-options)})]
+    (->MarathonApi http-client url (:spnego-auth http-options))))
+
+(defn agent-directory-content
+  "Lists files and directories contained in the path."
+  [{:keys [http-client spnego-auth]} host port directory]
+  (http-request http-client (str "http://" host ":" port "/files/browse")
+                :query-string {"path" directory}
+                :request-method :get
+                :spnego-auth spnego-auth
+                :throw-exceptions false))
+
+(defn agent-state
```

Comment:
For these `agent` functions, we're talking to Mesos, not Marathon.

Reply:
Renamed to `mesos-slave-directory-content` and `mesos-slave-state` to make it clearer that Mesos is involved.

Reply:
Mesos related tasks shouldn't require/be done through something designed to talk to Marathon.

Reply:
Created a separate `mesos-api` for mesos related tasks.